components/script/dom/customelementregistry.rs

```diff
@@ -0,0 +1,303 @@
+            element_definition_is_running: Cell::new(false),
+            definitions: DOMRefCell::new(HashMap::new()),
+        }
+    }
+
+    pub fn new(window: &Window) -> Root<CustomElementRegistry> {
+        reflect_dom_object(box CustomElementRegistry::new_inherited(window),
+                           window,
+                           CustomElementRegistryBinding::Wrap)
+    }
+
+    // Cleans up any active promises
+    // https://github.com/servo/servo/issues/15318
+    pub fn teardown(&self) {
+        self.when_defined.borrow_mut().clear()
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-customelementregistry-define
+    // Steps 10.1, 10.2
+    #[allow(unsafe_code)]
+    fn check_prototype(&self, constructor: HandleObject) -> ErrorResult {
+        let global_scope = self.window.upcast::<GlobalScope>();
+        rooted!(in(global_scope.get_cx()) let mut prototype = UndefinedValue());
+        unsafe {
+            // Step 10.1
+            let c_name = CString::new("prototype").unwrap();
+            if !JS_GetProperty(global_scope.get_cx(), constructor, c_name.as_ptr(), prototype.handle_mut()) {
+                return Err(Error::JSFailed);
+            }
+
+            // Step 10.2
+            if !prototype.is_object() {
+                return Err(Error::Type("constructor.protoype is not an object".to_owned()));
+            }
+        }
+        Ok(())
+    }
+}
+
+impl CustomElementRegistryMethods for CustomElementRegistry {
+    #[allow(unsafe_code, unrooted_must_root)]
+    // https://html.spec.whatwg.org/multipage/#dom-customelementregistry-define
+    fn Define(&self, name: DOMString, constructor_: Rc<Function>, options: &ElementDefinitionOptions) -> ErrorResult {
+        let global_scope = self.window.upcast::<GlobalScope>();
+        rooted!(in(global_scope.get_cx()) let constructor = constructor_.callback());
+
+        // Step 1
+        if unsafe { !IsConstructor(constructor.get()) } {
+            return Err(Error::Type("Second argument of CustomElementRegistry.defin is not a constructor".to_owned()));
+        }
+
+        // Step 2
+        if !is_valid_custom_element_name(&name) {
+            return Err(Error::Syntax)
+        }
+
+        // Step 3
+        if self.definitions.borrow().contains_key(&name) {
+            return Err(Error::NotSupported);
+        }
+
+        // Step 4
+        if self.definitions.borrow().iter().any(|(_, ref def)| def.constructor == constructor_) {
+            return Err(Error::NotSupported);
+        }
+
+        // Step 6
+        let extends = &options.extends;
+
+        // Steps 5, 7
+        let local_name = match *extends {
+            Some(ref extended_name) => {
+                // Step 7.1
+                if is_valid_custom_element_name(extended_name) {
+                    return Err(Error::NotSupported)
+                }
+
+                // TODO: 7.2
+                // Check that the element interface for extends is not HTMLUnknownElement
```

Comment:
I'm not entirely sure the best way to go about this. AFAICT we only determine the _element interface_ here: https://dxr.mozilla.org/servo/rev/779edd7c4aaf900f12fab378a378b0fc52d4c628/components/script/dom/create.rs#129-272.